File: c#-solution/0303. Range Sum Query - Immutable.cs
// 303. Range Sum Query - Immutable
// https://leetcode.com/problems/range-sum-query-immutable/description/
// Tag: Array, Prefix Sum, Easy


// 1. Brute Force
public class NumArray {

    new int [] arr;

    public NumArray(int[] nums) {
        arr = nums;
    }
    
    public int SumRange(int left, int right) {
        var sum = 0;
        for(int i = left; i<=right; i++){
            sum += arr[i];
        }
        return sum;
    }
}

// TC: O(n)
// SC: O(1)
// 2. Prefix Sum
public class NumArray {

    private int [] pre;

    public NumArray(int[] nums) { 
        pre = new int [nums.Length];
        pre[0] = nums[0];
        for(int i=1; i<nums.Length; i++){
            pre[i] = pre[i-1] + nums[i];
        }
    }
    
    public int SumRange(int left, int right) {
        if(left == 0){
            return pre[right];
        } else {
            return pre[right] - pre[left - 1];
        }
    }
}
// TC: O(n)
// SC: O(n)

File: problem/0303. Range Sum Query - Immutable.js
// 303. Range Sum Query - Immutable
// https://leetcode.com/problems/range-sum-query-immutable/
// Tag: Array, Prefix Sum, Easy


// brute force solution
// in ES5 syntax
/**
 * @param {number[]} nums
 */
var NumArray = function(nums) {
    this.arr = [];
    this.arr[0] = nums[0];
    for(let i=1; i<nums.length; i++){
        this.arr[i] = this.arr[i-1] + nums[i];
    }
};

/** 
 * @param {number} left 
 * @param {number} right
 * @return {number}
 */
NumArray.prototype.sumRange = function(left, right) {
    if(left === 0){
        return this.arr[right];
    } else {
        return this.arr[right] - this.arr[left -1];
    }
};

/** 
 * Your NumArray object will be instantiated and called as such:
 * var obj = new NumArray(nums)
 * var param_1 = obj.sumRange(left,right)
 */

// ES6 syntax
class NumArray {
    constructor(nums) {
        this.arr = [];
        this.arr[0] = nums[0];
        for (let i = 1; i < nums.length; i++) {
            this.arr[i] = this.arr[i - 1] + nums[i];
        }
    }

    sumRange(left, right) {
        return left === 0 ? this.arr[right] : this.arr[right] - this.arr[left - 1];
    }
}

// ES6 syntax with reduce for prefix sum
class NumArray { 
    constructor(nums) {
        this.prefixSum = nums.reduce((acc, cur, i) => {
            acc[i] = (i === 0)? cur : acc[i-1] + cur;
            return acc;
        }, []);
    }
    sumRange(left, right){
        return left === 0? this.prefixSum[right] : this.prefixSum[right] - this.prefixSum[left -1];
    }
};